osv/ecosystems/hackage.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2021 Google LLC
+# Copyright 2023 Fraser Tweedale
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Haskell ecosystem helpers.
+
+Contact the Haskell Security Response Team <security-advisories@haskell.org>
+if something is broken and you need help to fix it.
+
+"""
+
+import requests
+
+from . import config
+from .helper_base import Ecosystem, EnumerateError
+from .. import semver_index
+
+class Hackage(Ecosystem):
+  """Hackage (Haskell package index) ecosystem."""
+
+  _API_PACKAGE_URL = 'https://hackage.haskell.org/package/{package}.json'
+
+  def sort_key(self, version):
+    """Sort key.
+
+    The Haskell package version data type is defined at
+    https://hackage.haskell.org/package/Cabal-syntax/docs/Distribution-Types-Version.html
+
+    """
+    return [int(x) for x in version.split('.')]
+
+  def enumerate_versions(self,
+                         package,
+                         introduced,
+                         fixed=None,
+                         last_affected=None,
+                         limits=None):
+    """Enumerate versions."""
+    response = requests.get(
+      self._API_PACKAGE_URL.format(package=package), timeout=config.timeout)
+    if response.status_code == 404:
+      raise EnumerateError(f'Package {package} not found')
+    if response.status_code != 200:
+      raise RuntimeError(
+        f'Failed to get Hackage versions for {package} with: {response.text}'
+      )
+
+    response = response.json()
+    versions = list(resp.keys())
+
+    self.sort_versions(versions)
+    return self._get_affected_versions(versions, introduced, fixed,
+                                       last_affected, limits)
+
+
+class GHC(Ecosystem):
+  """Glasgow Haskell Compiler (GHC) ecosystem."""
+
+  _API_PACKAGE_URL = (
+    'https://gitlab.haskell.org'
+    '/api/v4/projects/3561/repository/tags?per_page=-1'
+  )
```

Comment:
It doesn't look like there's any tags for GHC versions before 7.2.1 in this repo. On the [GHC downloads page](https://www.haskell.org/ghc/download.html) I can see versions going all the way back to 0.29.

Is there a single source of truth for every GHC version ever released?

Reply:
Not surprising for a 30 year old project that underwent several VCS migrations.  Even the download page omits several earlier versions.  I asked around and the general consensus was Git tags would be the most reliable data source on an ongoing basis.  I will ask whether we could add tags for older versions.

Reply:
@michaelkedar adding tags for older versions is infeasible (because it is unknown/unknow*able* what commits particular releases were cut from).  What do you think about including a hardcoded list of the known historical public releases that don't have a corresponding Git tag?

Reply:
Yeah, a hardcoded list does seem like the only feasible way to have these historic versions. :+1: